src/Compilers/Core/Portable/EncodedStringText.cs

```diff
@@ -170,30 +170,32 @@ private static byte[] TryGetByteArrayFromStream(Stream data)
         }
 
         /// <summary>
-        /// If the MemoryStream was created with publiclyVisible=true, then we can access its buffer
+        /// If the MemoryStream was created with publiclyVisible=true, then we can try to access its buffer
         /// directly and save allocations in StreamReader. The input MemoryStream is not closed on exit.
         /// </summary>
         /// <returns>True if a byte array could be created.</returns>
         private static bool TryGetByteArrayFromMemoryStream(MemoryStream data, out byte[] buffer)
         {
             Debug.Assert(data.Position == 0);
 
-            try
+            if (PortableShim.MemoryStream.TryGetBuffer != null)
             {
-                if (PortableShim.MemoryStream.GetBuffer != null)
```

Comment:
@jaredpar I believe that not commonly used overload is exactly what was used by @nguerrera in the issue this PR is trying to fix (#12348).

What about this:

1. If `TryGetBuffer` is available (.Net Core and .Net 4.6+), use that (efficient and correct).
2. Otherwise (.Net 4.5.x), use `GetBuffer` (efficient, but not correct in the one edge case).

Though personally, I would prefer:

1. If `TryGetBuffer` is available (.Net Core and .Net 4.6+), use that (efficient and correct).
2. Otherwise (.Net 4.5.x), return `false` (inefficient, but correct).